packages/babel-helpers/src/index.js

```diff
@@ -1,10 +1,208 @@
+import traverse from "babel-traverse";
+import * as t from "babel-types";
 import helpers from "./helpers";
 
-export function get(name) {
-  const fn = helpers[name];
-  if (!fn) throw new ReferenceError(`Unknown helper ${name}`);
+function makePath(path) {
+  const parts = [];
 
-  return fn().expression;
+  for (; path; path = path.parentPath) {
+    parts.push(path.key);
+    if (path.inList) parts.push(path.listKey);
+  }
+
+  // Pop off the final "program" since these paths will be used from the Program NodePath.
+  parts.pop();
+
+  return parts.reverse().join(".");
+}
+
+/**
+ * Given a file AST for a given helper, get a bunch of metadata about it so that Babel can quickly render
+ * the helper is whatever context it is needed in.
+ */
+function getHelperMetadata(file) {
+  const globals = new Set();
+  const localBindingNames = new Set();
+
+  let exportName;
+  let exportPath;
+  const exportBindingAssignments = [];
+
+  traverse(file, {
+    ImportDeclaration() {
+      throw new Error();
+    },
+    ExportDefaultDeclaration(child) {
+      const decl = child.get("declaration");
+
+      if (decl.isFunctionDeclaration()) {
+        if (!decl.node.id) {
+          throw new Error(
+            "Helpers should give names to their exported func declaration",
+          );
+        }
+
+        exportName = decl.node.id.name;
+      }
+      exportPath = makePath(child);
+    },
+    ExportAllDeclaration(child) {
+      throw child.buildCodeFrameError("Helpers can only export default");
+    },
+    ExportNamedDeclaration(child) {
+      throw child.buildCodeFrameError("Helpers can only export default");
+    },
+  });
+
+  traverse(file, {
```

Comment:
Why are we doing multiple traversals?

Reply:
The helpers are such tiny ASTs that it didn't seem like it would be an issue to traverse multiple times, especially since helpers are memoized, so at most this is called once per-helper-per-file.